File: web/src/vision/quad-detector.ts
/**
 * Document quad detection — docs/ALGORITHM.md steps 1–8 on an OpenCV Mat.
 *
 * The caller supplies RGBA pixels (ImageData drawn from the <video> at
 * processing size); this module performs grayscale conversion, downscale,
 * blur, adaptive Canny, dilate, contour extraction and all quad acceptance
 * checks. All detection constants live in DEFAULT_DETECTION_PARAMS
 * (`DetectionParams`, mirrored by `DetectionConfig` in the Flutter SDK).
 */
import type { CV, CvMat, ImageDataLike } from '../core/opencv-loader';
import { aspectAccepted, aspectRatio, quadArea, type Point, type Quad } from './geometry';

/**
 * Extract a document's 4 corners from a contour as the farthest point from
 * the contour centroid within each quadrant (jscanify, MIT — see NOTICE).
 * Returns corners already ordered TL, TR, BR, BL, or null if any quadrant
 * is empty (not a quadrilateral-ish blob).
 */
function extremeCornerQuad(contour: CvMat): Quad | null {
  const data = contour.data32S;
  const n = data.length / 2;
  if (n < 4) return null;
  let cx = 0;
  let cy = 0;
  for (let i = 0; i < n; i++) {
    cx += data[i * 2];
    cy += data[i * 2 + 1];
  }
  cx /= n;
  cy /= n;

  let tl: Point | null = null;
  let tr: Point | null = null;
  let br: Point | null = null;
  let bl: Point | null = null;
  let dtl = -1;
  let dtr = -1;
  let dbr = -1;
  let dbl = -1;
  for (let i = 0; i < n; i++) {
    const x = data[i * 2];
    const y = data[i * 2 + 1];
    const dx = x - cx;
    const dy = y - cy;
    const d = dx * dx + dy * dy;
    if (dx <= 0 && dy <= 0) {
      if (d > dtl) {
        dtl = d;
        tl = { x, y };
      }
    } else if (dx > 0 && dy <= 0) {
      if (d > dtr) {
        dtr = d;
        tr = { x, y };
      }
    } else if (dx > 0 && dy > 0) {
      if (d > dbr) {
        dbr = d;
        br = { x, y };
      }
    } else if (d > dbl) {
      dbl = d;
      bl = { x, y };
    }
  }
  if (!tl || !tr || !br || !bl) return null;
  return [tl, tr, br, bl];
}

/** All pipeline constants (docs/ALGORITHM.md). One object per platform. */
export interface DetectionParams {
  /** Longest processed dimension (step 2). */
  processingMaxDim: number;
  /** Gaussian blur kernel size (step 3). */
  gaussianKernelSize: number;
  /** Adaptive Canny clamp range (step 4). */
  cannyClampMin: number;
  cannyClampMax: number;
  /** Canny thresholds = [lowerFactor, upperFactor] x median. */
  cannyLowerFactor: number;
  cannyUpperFactor: number;
  /** Morphological-close kernel size (step 5) — bridges card-edge gaps. */
  closeKernelSize: number;
  /** Top-N largest contours to examine (step 6). */
  maxContourCandidates: number;
  /** Contour area >= this fraction of the guide-rect area (step 7). */
  minGuideAreaFrac: number;
  /** Detected corners may sit this fraction of the guide beyond its edges. */
  guideCornerMarginFrac: number;
  /** Corners must be >= this many px inside the processed frame border. */
  borderMarginPx: number;
  /** Aspect tolerance vs the document target (step 8). */
  aspectTolerance: number;
  /** Quad area within [min, max] x guide area (step 8). */
  guideAreaMinFrac: number;
  guideAreaMaxFrac: number;
  /** Stability window (step 9). */
  stabilityWindow: number;
  minStableFrames: number;
  maxCornerDriftFrac: number;
  /** Laplacian-variance sharpness threshold (step 10). */
  minSharpness: number;
  /** Ring buffer of accepted frames (step 10). */
  ringBufferSize: number;
  /** Manual capture button appears after this many ms (UX section). */
  manualFallbackMs: number;
  /** Frame budget: process at most this many frames per second. */
  targetFps: number;
  /** Guide rect width as a fraction of the processed frame width. */
  guideWidthFrac: number;
  /**
   * Aspect window for the `cardLike` signal — a landscape rectangle. Wide
   * enough for ID-1 (1.586) and passport (1.42) with slack, tight enough
   * to exclude a near-square/portrait face outline.
   */
  cardLikeAspectMin: number;
  cardLikeAspectMax: number;
  /** EMA factor for corner smoothing (0..1; lower = steadier). */
  cornerSmoothingAlpha: number;
  /** Corner move beyond this many px snaps to raw (fast reposition). */
  cornerSmoothingResetPx: number;
  /**
   * "Easy" occupancy snap (web capture model). The guide is OCCUPIED by a
   * detailed object when the fraction of Canny-edge pixels inside the guide
   * rect (`guideEdgeDensity`) is >= this. A text-filled card gives a high
   * density; an empty wall is ~0. This lets auto-capture fire without the
   * fragile edge-contour quad being accepted.
   */
  occupancyMinEdgeDensity: number;
  /**
   * Easy snap: the mean absolute difference of the gray guide crop vs the
   * previous frame's crop must be <= this to count as motion-stable.
   */
  easyMotionMaxMeanDiff: number;
  /**
   * Easy snap: consecutive occupied + motion-stable + sharp frames required
   * before firing (~0.3 s at the default frame rate).
   */
  easyStableFrames: number;
}

export const DEFAULT_DETECTION_PARAMS: DetectionParams = {
  processingMaxDim: 480,
  gaussianKernelSize: 5,
  cannyClampMin: 30,
  cannyClampMax: 200,
  cannyLowerFactor: 0.66,
  cannyUpperFactor: 1.33,
  // 7: a wider close kernel bridges glare/low-contrast gaps in the card
  // border so it survives as one closed contour.
  closeKernelSize: 7,
  maxContourCandidates: 8,
  minGuideAreaFrac: 0.45,
  guideCornerMarginFrac: 0.12,
  borderMarginPx: 6,
  aspectTolerance: 0.3,
  guideAreaMinFrac: 0.5,
  // 1.35: users naturally overfill the guide a little.
  guideAreaMaxFrac: 1.35,
  // Corners are EMA-smoothed before this check, so hand tremor is damped
  // and the trigger is easy to reach: 2-of-4 frames at 9% drift (~0.2 s).
  stabilityWindow: 4,
  minStableFrames: 2,
  maxCornerDriftFrac: 0.09,
  // Webcams are soft; the ring buffer still submits the SHARPEST frame.
  minSharpness: 30,
  ringBufferSize: 6,
  // Auto-capture only fires on an aligned document quad; surface the manual
  // button quickly so a user is never stuck if the quad won't lock.
  manualFallbackMs: 4_000,
  targetFps: 12,
  guideWidthFrac: 0.8,
  cardLikeAspectMin: 1.25,
  cardLikeAspectMax: 2.4,
  // EMA smoothing: 0.45 damps jitter while still tracking real movement;
  // a >60px jump snaps to raw so fast repositions aren't laggy.
  cornerSmoothingAlpha: 0.45,
  cornerSmoothingResetPx: 60,
  // Easy occupancy snap: a text-filled card fills the guide with edges
  // (density well above 0.03); an empty wall stays near 0. Motion <= 8
  // mean-abs-diff is "held steady"; 4 such frames (~0.3 s) fires capture.
  occupancyMinEdgeDensity: 0.03,
  easyMotionMaxMeanDiff: 8,
  easyStableFrames: 4,
};

export interface GuideRect {
  x: number;
  y: number;
  width: number;
  height: number;
}

/** Why the best candidate was rejected — drives the UX state machine. */
export type RejectReason = 'noQuad' | 'moveCloser' | 'alignCard';

export interface QuadDetectionResult {
  /** Accepted quad (TL,TR,BR,BL) in PROCESSED-frame coordinates, or null. */
  quad: Quad | null;
  reason: RejectReason | null;
  /**
   * A substantial, roughly card-shaped rectangle was found this frame —
   * a 4-point convex contour (or hull) that filled a large part of the
   * guide with a landscape aspect, EVEN IF it was rejected for
   * aspect/centroid/oversize. This is the "a document-like object is
   * present" signal; a face or empty scene is never cardLike, so assisted
   * capture keys off it (not raw pixel motion).
   */
  cardLike: boolean;
  /**
   * Fraction (0..1) of Canny-edge pixels inside the guide rect this frame —
   * the "guide is occupied by a detailed object" signal. A text-filled card
   * scores high; an empty wall scores ~0. Drives the easy occupancy snap
   * (document-capture) independently of quad acceptance.
   */
  guideEdgeDensity: number;
  /**
   * Processed grayscale mat (post-downscale, pre-blur) for sharpness
   * scoring. THE CALLER MUST CALL `.delete()` on it.
   */
  gray: CvMat;
  /** Multiply processed coords by this to get source-image coords. */
  scaleBack: number;
  processedWidth: number;
  processedHeight: number;
}

/** Centered guide rect matching the document aspect (UX + step 8 checks). */
export function computeGuideRect(
  frameWidth: number,
  frameHeight: number,
  targetAspect: number,
  widthFrac: number = DEFAULT_DETECTION_PARAMS.guideWidthFrac,
): GuideRect {
  let gw = frameWidth * widthFrac;
  let gh = gw / targetAspect;
  const maxH = frameHeight * 0.8;
  if (gh > maxH) {
    gh = maxH;
    gw = gh * targetAspect;
  }
  return { x: (frameWidth - gw) / 2, y: (frameHeight - gh) / 2, width: gw, height: gh };
}

/** Median of an 8-bit grayscale image via its histogram. */
function grayscaleMedian(gray: CvMat): number {
  const hist = new Uint32Array(256);
  const data = gray.data;
  for (let i = 0; i < data.length; i++) hist[data[i]]++;
  const half = data.length / 2;
  let cumulative = 0;
  for (let v = 0; v < 256; v++) {
    cumulative += hist[v];
    if (cumulative >= half) return v;
  }
  return 255;
}

function clamp(v: number, lo: number, hi: number): number {
  return Math.min(hi, Math.max(lo, v));
}

/**
 * Run detection steps 1–8 on one RGBA frame.
 *
 * @param cv          initialized OpenCV module (loadOpenCv()).
 * @param imageData   RGBA pixels; ideally already at processing size
 *                    (longest side <= params.processingMaxDim). Larger
 *                    inputs are downscaled here (step 2) and `scaleBack`
 *                    reports the processed -> source factor.
 * @param targetAspect 1.586 (ID-1) or 1.42 (passport).
 * @param guide       guide rect in PROCESSED coordinates; pass null to
 *                    derive a centered default from the aspect.
 */
export function detectQuad(
  cv: CV,
  imageData: ImageDataLike,
  targetAspect: number,
  guide: GuideRect | null = null,
  params: DetectionParams = DEFAULT_DETECTION_PARAMS,
): QuadDetectionResult {
  const rgba = cv.matFromImageData(imageData);
  let gray = new cv.Mat();
  cv.cvtColor(rgba, gray, cv.COLOR_RGBA2GRAY);
  rgba.delete();

  // Step 2: downscale so the longest dimension == processingMaxDim.
  let scaleBack = 1;
  const maxDim = Math.max(gray.cols, gray.rows);
  if (maxDim > params.processingMaxDim) {
    scaleBack = maxDim / params.processingMaxDim;
    const w = Math.round(gray.cols / scaleBack);
    const h = Math.round(gray.rows / scaleBack);
    const resized = new cv.Mat();
    cv.resize(gray, resized, new cv.Size(w, h), 0, 0, cv.INTER_AREA);
    gray.delete();
    gray = resized;
  }
  const width = gray.cols;
  const height = gray.rows;
  const guideRect = guide ?? computeGuideRect(width, height, targetAspect, params.guideWidthFrac);
  const guideArea = guideRect.width * guideRect.height;

  // Step 3: Gaussian blur (on a working copy; `gray` is returned for
  // sharpness scoring and must stay unblurred).
  const blurred = new cv.Mat();
  cv.GaussianBlur(
    gray,
    blurred,
    new cv.Size(params.gaussianKernelSize, params.gaussianKernelSize),
    0,
  );

  // Step 4: adaptive Canny from the grayscale median, clamped to [30, 200].
  const median = grayscaleMedian(blurred);
  const lower = clamp(params.cannyLowerFactor * median, params.cannyClampMin, params.cannyClampMax);
  const upper = clamp(params.cannyUpperFactor * median, params.cannyClampMin, params.cannyClampMax);
  const edges = new cv.Mat();
  cv.Canny(blurred, edges, lower, upper);
  blurred.delete();

  // Occupancy signal: fraction of Canny-edge pixels inside the guide rect
  // (computed on `edges` BEFORE it is consumed/deleted). meanStdDev of a
  // binary edge map is 255 * (edge fraction), so mean/255 = density.
  let guideEdgeDensity = 0;
  {
    const gx = clamp(Math.floor(guideRect.x), 0, Math.max(0, width - 1));
    const gy = clamp(Math.floor(guideRect.y), 0, Math.max(0, height - 1));
    const gw = clamp(Math.round(guideRect.width), 1, width - gx);
    const gh = clamp(Math.round(guideRect.height), 1, height - gy);
    const roi = edges.roi(new cv.Rect(gx, gy, gw, gh));
    const mean = new cv.Mat();
    const stddev = new cv.Mat();
    try {
      cv.meanStdDev(roi, mean, stddev);
      guideEdgeDensity = mean.data64F[0] / 255;
    } finally {
      roi.delete();
      mean.delete();
      stddev.delete();
    }
  }

  // Step 5: morphological CLOSE with a larger kernel to bridge gaps in the
  // card's edge (glare, low contrast, finger occlusion) so the border
  // survives as ONE closed contour — the key to robust detection.
  const kernel = cv.getStructuringElement(
    cv.MORPH_RECT,
    new cv.Size(params.closeKernelSize, params.closeKernelSize),
  );
  const closed = new cv.Mat();
  cv.morphologyEx(edges, closed, cv.MORPH_CLOSE, kernel);
  edges.delete();
  kernel.delete();

  // Step 6: external contours, largest first.
  const contours = new cv.MatVector();
  const hierarchy = new cv.Mat();
  cv.findContours(closed, contours, hierarchy, cv.RETR_EXTERNAL, cv.CHAIN_APPROX_SIMPLE);
  closed.delete();
  hierarchy.delete();

  const indexed: Array<{ index: number; area: number }> = [];
  for (let i = 0; i < contours.size(); i++) {
    indexed.push({ index: i, area: cv.contourArea(contours.get(i)) });
  }
  indexed.sort((a, b) => b.area - a.area);

  let accepted: Quad | null = null;
  let reason: RejectReason = 'noQuad';
  let cardLike = false;

  // Step 7 (jscanify approach, MIT — see NOTICE): the card is the largest
  // contour that fills the guide. Its 4 corners are the points farthest
  // from the contour's centroid in each quadrant — robust to rounded
  // corners, broken edges and fingers, unlike a strict approxPolyDP that
  // demands exactly 4 convex vertices.
  const marginX = guideRect.width * params.guideCornerMarginFrac;
  const marginY = guideRect.height * params.guideCornerMarginFrac;

  for (const { index, area: contourArea } of indexed.slice(0, params.maxContourCandidates)) {
    if (contourArea < params.minGuideAreaFrac * guideArea) break; // sorted desc → rest smaller
    const quad = extremeCornerQuad(contours.get(index));
    if (!quad) continue;

    const area = quadArea(quad);
    const aspect = aspectRatio(quad);
    // Orientation-agnostic card-like signal (UX hint only): a substantial
    // rectangle whose long/short ratio is landscape-ish. A near-round face
    // blob (ratio ~1) is excluded; a portrait passport (0.71 → 1.41) counts.
    const ratio = aspect >= 1 ? aspect : 1 / aspect;
    if (
      area >= params.minGuideAreaFrac * guideArea &&
      ratio >= params.cardLikeAspectMin &&
      ratio <= params.cardLikeAspectMax
    ) {
      cardLike = true;
    }

    // Corners must sit within the guide (+ margin) and inside the frame.
    const withinGuide = quad.every(
      (p) =>
        p.x >= guideRect.x - marginX &&
        p.x <= guideRect.x + guideRect.width + marginX &&
        p.y >= guideRect.y - marginY &&
        p.y <= guideRect.y + guideRect.height + marginY,
    );
    const inBorder = quad.every(
      (p) =>
        p.x >= params.borderMarginPx &&
        p.y >= params.borderMarginPx &&
        p.x <= width - params.borderMarginPx &&
        p.y <= height - params.borderMarginPx,
    );
    if (!withinGuide || !inBorder) {
      reason = 'alignCard';
      continue;
    }
    if (area < params.guideAreaMinFrac * guideArea) {
      reason = 'moveCloser';
      continue;
    }
    if (area > params.guideAreaMaxFrac * guideArea) {
      reason = 'alignCard';
      continue;
    }
    if (!aspectAccepted(aspect, targetAspect, params.aspectTolerance)) {
      reason = 'alignCard';
      continue;
    }
    accepted = quad;
    break;
  }
  contours.delete();

  return {
    quad: accepted,
    reason: accepted ? null : reason,
    cardLike: accepted !== null || cardLike,
    guideEdgeDensity,
    gray,
    scaleBack,
    processedWidth: width,
    processedHeight: height,
  };
}

File: flutter/lib/src/vision/quad_detector.dart
import 'dart:math' as math;
import 'dart:typed_data';

import 'package:opencv_dart/opencv_dart.dart' as cv;

import 'blur_scorer.dart';
import 'quad_geometry.dart' as geo;

/// All tunable constants of the detection pipeline, with the defaults
/// mandated by docs/ALGORITHM.md. The Web SDK's `DetectionParams` carries
/// identical values.
class DetectionConfig {
  /// Longest dimension of the processing frame (step 2).
  final int processingMaxDim;

  /// Canny threshold clamp range (step 4).
  final double cannyClampLow;
  final double cannyClampHigh;

  /// Contours examined per frame, sorted by area descending (step 6).
  final int maxContourCandidates;

  /// approxPolyDP epsilon as a fraction of arc length (step 7).
  final double approxEpsilonFrac;

  /// Minimum contour area as a fraction of the processed frame (step 7).
  final double minFrameAreaFrac;

  /// Minimum contour area as a fraction of the guide rect (step 7).
  final double minGuideAreaFrac;

  /// Corners must be at least this many px inside the processed frame.
  final double borderMarginPx;

  /// Aspect-ratio acceptance tolerance (step 8).
  final double aspectTolerance;

  /// Guide-alignment window: quad area vs guide area (step 8).
  final double guideAreaMinFrac;
  final double guideAreaMaxFrac;

  /// Stability tracking (step 9).
  final int stabilityWindow;
  final int minStableFrames;
  final double maxCornerDriftFrac;

  /// Laplacian-variance sharpness threshold (step 10).
  final double minSharpness;

  /// Accepted-frame ring buffer length (step 10).
  final int frameBufferSize;

  /// Manual capture button appears after this long without auto-capture.
  final int manualFallbackMs;

  /// Assisted fallback: if no quad has been accepted after this many ms,
  /// auto-capture switches to guide-region mode — fingers over the card
  /// edge routinely break contour-based quad detection, and users should
  /// not need the manual button for that.
  final int assistedFallbackMs;

  /// Consecutive sharp+stable guide frames required in assisted mode.
  final int assistedStableFrames;

  /// Max mean abs pixel diff (0–255) between guide crops to count stable.
  final double assistedMaxMeanDiff;

  /// Frame-processing budget (frames per second).
  final int maxProcessingFps;

  const DetectionConfig({
    this.processingMaxDim = 480,
    this.cannyClampLow = 30,
    this.cannyClampHigh = 200,
    this.maxContourCandidates = 5,
    this.approxEpsilonFrac = 0.02,
    this.minFrameAreaFrac = 0.08,
    this.minGuideAreaFrac = 0.5,
    this.borderMarginPx = 8,
    this.aspectTolerance = 0.25,
    this.guideAreaMinFrac = 0.6,
    // 1.3: users naturally overfill the guide a little; 1.15 rejected that.
    this.guideAreaMaxFrac = 1.3,
    // Handheld reality: a card held in front of a camera always tremors a
    // few px and acceptance flickers between direct/hull corners, so the
    // trigger is 4-of-6 frames with 3.5% drift — ~0.5 s of a normal hold.
    this.stabilityWindow = 6,
    this.minStableFrames = 4,
    this.maxCornerDriftFrac = 0.035,
    // Cameras are soft; the frame buffer still submits the SHARPEST frame,
    // and 60 comfortably rejects genuine motion blur.
    this.minSharpness = 60,
    this.frameBufferSize = 5,
    this.manualFallbackMs = 10000,
    this.assistedFallbackMs = 3000,
    this.assistedStableFrames = 4,
    this.assistedMaxMeanDiff = 10,
    this.maxProcessingFps = 10,
  });
}

/// Why a frame did not produce an accepted quad (drives the UX state).
enum QuadStatus {
  /// Accepted quad.
  found,

  /// No plausible quad in frame.
  notFound,

  /// Quad found but too small vs the guide (area < 60% of guide).
  moveCloser,

  /// Quad found but aspect/centroid/size checks failed.
  alignCard,
}

/// Outcome of one processed frame.
class QuadDetectionResult {
  final QuadStatus status;

  /// Ordered corners TL, TR, BR, BL in SOURCE image coordinates
  /// (only when [status] == [QuadStatus.found]).
  final List<math.Point<double>>? corners;

  /// Aspect ratio of the accepted quad.
  final double? aspect;

  /// Laplacian variance of the quad's bounding-box crop (0 if not found).
  final double sharpness;

  const QuadDetectionResult._(
    this.status, {
    this.corners,
    this.aspect,
    this.sharpness = 0,
  });

  bool get isFound => status == QuadStatus.found;

  bool isSharp(DetectionConfig config) => sharpness >= config.minSharpness;
}

/// OpenCV implementation of docs/ALGORITHM.md steps 1–8 (+10 sharpness).
class QuadDetector {
  final DetectionConfig config;

  QuadDetector({this.config = const DetectionConfig()});

  /// Builds a grayscale Mat from the Y plane of an NV21/YUV420 camera
  /// image, respecting [strideBytes] (`bytesPerRow`).
  static cv.Mat grayFromNv21Y(
    Uint8List yPlane,
    int width,
    int height, {
    int? strideBytes,
  }) {
    final stride = strideBytes ?? width;
    if (stride == width) {
      final mat = cv.Mat.fromList(
        height,
        width,
        cv.MatType.CV_8UC1,
        yPlane.length == width * height
            ? yPlane
            : Uint8List.sublistView(yPlane, 0, width * height),
      );
      return mat;
    }
    // Strip the row padding.
    final packed = Uint8List(width * height);
    for (var row = 0; row < height; row++) {
      packed.setRange(
        row * width,
        (row + 1) * width,
        Uint8List.sublistView(yPlane, row * stride, row * stride + width),
      );
    }
    return cv.Mat.fromList(height, width, cv.MatType.CV_8UC1, packed);
  }

  /// Builds a grayscale Mat from BGRA8888 bytes (iOS camera stream).
  static cv.Mat grayFromBgra(
    Uint8List bgra,
    int width,
    int height, {
    int? strideBytes,
  }) {
    final stride = strideBytes ?? width * 4;
    Uint8List packed;
    if (stride == width * 4) {
      packed = bgra.length == width * height * 4
          ? bgra
          : Uint8List.sublistView(bgra, 0, width * height * 4);
    } else {
      packed = Uint8List(width * height * 4);
      for (var row = 0; row < height; row++) {
        packed.setRange(
          row * width * 4,
          (row + 1) * width * 4,
          Uint8List.sublistView(bgra, row * stride, row * stride + width * 4),
        );
      }
    }
    final bgraMat = cv.Mat.fromList(height, width, cv.MatType.CV_8UC4, packed);
    try {
      return cv.cvtColor(bgraMat, cv.COLOR_BGRA2GRAY);
    } finally {
      bgraMat.dispose();
    }
  }

  /// Builds a BGR Mat from a full NV21 buffer (Y + interleaved VU), used
  /// for the stream-frame capture fallback.
  static cv.Mat bgrFromNv21(Uint8List nv21, int width, int height) {
    final yuv = cv.Mat.fromList(
      height * 3 ~/ 2,
      width,
      cv.MatType.CV_8UC1,
      nv21.length == width * height * 3 ~/ 2
          ? nv21
          : Uint8List.sublistView(nv21, 0, width * height * 3 ~/ 2),
    );
    try {
      return cv.cvtColor(yuv, cv.COLOR_YUV2BGR_NV21);
    } finally {
      yuv.dispose();
    }
  }

  /// Rotates a Mat so the image is upright. [rotationDegrees] is the
  /// clockwise rotation needed (0/90/180/270), e.g. the camera sensor
  /// orientation on Android.
  static cv.Mat rotateUpright(cv.Mat src, int rotationDegrees) {
    switch (rotationDegrees % 360) {
      case 90:
        return cv.rotate(src, cv.ROTATE_90_CLOCKWISE);
      case 180:
        return cv.rotate(src, cv.ROTATE_180);
      case 270:
        return cv.rotate(src, cv.ROTATE_90_COUNTERCLOCKWISE);
      default:
        return src.clone();
    }
  }

  /// Runs steps 2–8 (+ sharpness) on an upright grayscale frame.
  ///
  /// [guideRect] is the on-screen guide rectangle mapped into SOURCE image
  /// coordinates. [targetAspect] is 1.586 (ID-1) or 1.42 (passport).
  /// Returned corners are in SOURCE image coordinates.
  QuadDetectionResult detect(
    cv.Mat gray, {
    required math.Rectangle<double> guideRect,
    required double targetAspect,
  }) {
    final srcW = gray.cols, srcH = gray.rows;
    final maxDim = math.max(srcW, srcH);
    final scale = maxDim > config.processingMaxDim
        ? config.processingMaxDim / maxDim
        : 1.0;
    final procW = (srcW * scale).round();
    final procH = (srcH * scale).round();

    final mats = <cv.Mat>[];
    try {
      cv.Mat proc;
      if (scale < 1.0) {
        proc = cv.resize(gray, (procW, procH), interpolation: cv.INTER_AREA);
        mats.add(proc);
      } else {
        proc = gray;
      }

      // Step 3: Gaussian blur 5×5, sigma 0.
      final blurred = cv.gaussianBlur(proc, (5, 5), 0);
      mats.add(blurred);

      // Step 4: adaptive Canny from the grayscale median.
      final m = _median(blurred);
      final lower = (0.66 * m).clamp(
        config.cannyClampLow,
        config.cannyClampHigh,
      );
      final upper = (1.33 * m).clamp(
        config.cannyClampLow,
        config.cannyClampHigh,
      );
      final edges = cv.canny(blurred, lower, upper);
      mats.add(edges);

      // Step 5: dilate 3×3 rect, 1 iteration.
      final kernel = cv.getStructuringElement(cv.MORPH_RECT, (3, 3));
      mats.add(kernel);
      final dilated = cv.dilate(edges, kernel, iterations: 1);
      mats.add(dilated);

      // Step 6: external contours, top candidates by area.
      final (contours, hierarchy) = cv.findContours(
        dilated,
        cv.RETR_EXTERNAL,
        cv.CHAIN_APPROX_SIMPLE,
      );
      hierarchy.dispose();
      try {
        final ranked = <(double, cv.VecPoint)>[
          for (final c in contours) (cv.contourArea(c), c),
        ]..sort((a, b) => b.$1.compareTo(a.$1));

        final guideProc = math.Rectangle<double>(
          guideRect.left * scale,
          guideRect.top * scale,
          guideRect.width * scale,
          guideRect.height * scale,
        );
        final guideArea = guideProc.width * guideProc.height;
        final frameArea = (procW * procH).toDouble();

        var rejection = QuadStatus.notFound;
        final limit = math.min(config.maxContourCandidates, ranked.length);

        for (var i = 0; i < limit; i++) {
          final contour = ranked[i].$2;
          // Step 7: quadrilateral test.
          var pts = _fourPointApprox(contour);
          if (pts == null) {
            // Fingers holding a card break its outline into >4 vertices;
            // the convex hull smooths those intrusions back into a
            // quadrilateral.
            final hullMat = cv.convexHull(contour);
            final hull = cv.VecPoint.fromMat(hullMat);
            try {
              pts = _fourPointApprox(hull);
            } finally {
              hull.dispose();
              hullMat.dispose();
            }
          }
          if (pts == null) continue;

          final ordered = geo.orderCorners(pts);
          if (!geo.anglesOk(ordered)) continue;

          final area = geo.quadArea(ordered);
          if (area < config.minFrameAreaFrac * frameArea) continue;
          if (area < config.minGuideAreaFrac * guideArea) continue;

          final insideBorder = ordered.every(
            (p) =>
                p.x >= config.borderMarginPx &&
                p.y >= config.borderMarginPx &&
                p.x <= procW - config.borderMarginPx &&
                p.y <= procH - config.borderMarginPx,
          );
          if (!insideBorder) continue;

          // Step 8: shape + guide-alignment checks.
          final aspect = geo.aspectRatio(ordered);
          if (!geo.aspectAccepted(
            aspect,
            targetAspect,
            tolerance: config.aspectTolerance,
          )) {
            rejection = QuadStatus.alignCard;
            continue;
          }

          final centroid = geo.quadCentroid(ordered);
          if (!guideProc.containsPoint(
            math.Point<double>(centroid.x, centroid.y),
          )) {
            rejection = QuadStatus.alignCard;
            continue;
          }

          if (area < config.guideAreaMinFrac * guideArea) {
            rejection = QuadStatus.moveCloser;
            continue;
          }
          if (area > config.guideAreaMaxFrac * guideArea) {
            rejection = QuadStatus.alignCard;
            continue;
          }

          // Step 10: sharpness on the quad's bounding-box crop of the
          // processed grayscale.
          final sharpness = _quadSharpness(proc, ordered, procW, procH);

          final sourceCorners = [
            for (final p in ordered)
              math.Point<double>(p.x / scale, p.y / scale),
          ];
          return QuadDetectionResult._(
            QuadStatus.found,
            corners: sourceCorners,
            aspect: aspect,
            sharpness: sharpness,
          );
        }
        return QuadDetectionResult._(rejection);
      } finally {
        contours.dispose();
      }
    } finally {
      for (final mat in mats) {
        mat.dispose();
      }
    }
  }

  /// Step 7 polygon approximation: `approxPolyDP` with
  /// ε = [DetectionConfig.approxEpsilonFrac] × arcLength. Returns the four
  /// corner points iff the approximation has exactly 4 points and is
  /// convex, else null (mirrors the Web SDK's `fourPointApprox`).
  List<math.Point<double>>? _fourPointApprox(cv.VecPoint shape) {
    final eps = config.approxEpsilonFrac * cv.arcLength(shape, true);
    final approx = cv.approxPolyDP(shape, eps, true);
    try {
      if (approx.length != 4) return null;
      if (!cv.isContourConvex(approx)) return null;
      return [
        for (final p in approx)
          math.Point<double>(p.x.toDouble(), p.y.toDouble()),
      ];
    } finally {
      approx.dispose();
    }
  }

  double _quadSharpness(
    cv.Mat proc,
    List<math.Point<double>> ordered,
    int procW,
    int procH,
  ) {
    var minX = double.infinity, minY = double.infinity;
    var maxX = -double.infinity, maxY = -double.infinity;
    for (final p in ordered) {
      minX = math.min(minX, p.x);
      minY = math.min(minY, p.y);
      maxX = math.max(maxX, p.x);
      maxY = math.max(maxY, p.y);
    }
    final x = minX.floor().clamp(0, procW - 2);
    final y = minY.floor().clamp(0, procH - 2);
    final w = (maxX.ceil() - x).clamp(1, procW - x);
    final h = (maxY.ceil() - y).clamp(1, procH - y);
    final roi = proc.region(cv.Rect(x, y, w, h));
    try {
      return laplacianVariance(roi);
    } finally {
      roi.dispose();
    }
  }

  /// Median gray level of a CV_8UC1 Mat via a 256-bin histogram.
  static double _median(cv.Mat gray8u) {
    final bytes = gray8u.data;
    final hist = List<int>.filled(256, 0);
    for (final b in bytes) {
      hist[b]++;
    }
    final half = bytes.length ~/ 2;
    var cumulative = 0;
    for (var v = 0; v < 256; v++) {
      cumulative += hist[v];
      if (cumulative > half) return v.toDouble();
    }
    return 255;
  }
}
